azure/scope/identity.go

```diff
@@ -132,15 +140,62 @@ func NewManagedControlPlaneCredentialsProvider(ctx context.Context, kubeClient c
 }
 
 // GetAuthorizer returns an Azure authorizer based on the provided azure identity. It delegates to AzureCredentialsProvider with AzureManagedControlPlane metadata.
-func (p *ManagedControlPlaneCredentialsProvider) GetAuthorizer(ctx context.Context, resourceManagerEndpoint string) (autorest.Authorizer, error) {
-	return p.AzureCredentialsProvider.GetAuthorizer(ctx, resourceManagerEndpoint, p.AzureManagedControlPlane.ObjectMeta)
+func (p *ManagedControlPlaneCredentialsProvider) GetAuthorizer(ctx context.Context, input GetAuthorizerInput) (autorest.Authorizer, error) {
+	return p.AzureCredentialsProvider.GetAuthorizer(ctx, input, p.AzureManagedControlPlane.ObjectMeta)
 }
 
 // GetAuthorizer returns an Azure authorizer based on the provided azure identity and cluster metadata.
-func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context, resourceManagerEndpoint string, clusterMeta metav1.ObjectMeta) (autorest.Authorizer, error) {
-	azureIdentityType, err := getAzureIdentityType(p.Identity)
+func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context, input GetAuthorizerInput, clusterMeta metav1.ObjectMeta) (autorest.Authorizer, error) {
+	var spt *adal.ServicePrincipalToken
+	switch p.Identity.Spec.Type {
+	case infrav1.ServicePrincipal:
+		if err := createAzureIdentityWithBindings(ctx, p.Identity, clusterMeta, p.Client); err != nil {
+			return nil, err
+		}
+
+		msiEndpoint, err := adal.GetMSIVMEndpoint()
+		if err != nil {
+			return nil, errors.Errorf("failed to get MSI endpoint: %v", err)
+		}
+
+		spt, err = adal.NewServicePrincipalTokenFromMSIWithUserAssignedID(msiEndpoint, input.resourceManagerEndpoint, p.Identity.Spec.ClientID)
+		if err != nil {
+			return nil, errors.Errorf("failed to get token from service principal identity: %v", err)
+		}
+
+	case infrav1.ManualServicePrincipal:
+		oauthConfig, err := adal.NewOAuthConfig(input.aadEndpoint, input.tenantID)
+		if err != nil {
+			return nil, err
+		}
+
+		var clientSecret corev1.Secret
+		if err := p.Client.Get(ctx, client.ObjectKey{Namespace: p.Identity.Spec.ClientSecret.Namespace, Name: p.Identity.Spec.ClientSecret.Name}, &clientSecret); err != nil {
+			return nil, err
+		}
+
+		clientSecretVal, ok := clientSecret.StringData["clientSecret"]
```

Comment:
I see some reusable code with #1523 👀

Reply:
Yes, many of the things from #1523 can be used here. I'll  refactor it once it's merged.